js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts

```diff
@@ -0,0 +1,242 @@
+     * Function to resolve a list of choices based on the provided params.
+     *
+     * What this function is passed depends on the implementation but when used with
+     * [useAsyncChoices](doc:useAsyncChoices) it will be passed `value` (the currently
+     * selected value), `query` (the query object eg. to filter results with), `paginator` (the current
+     * paginator if any) and `listOptions` (any additional options passed on the `listOptions` prop to `useAsyncChoices`).
+     * Official presto widgets all use `useAsyncChoices` and so will use these parameters.
+     */
+    list(params: Record<string, any>): Promise<ItemType[]>;
+
+    /**
+     * Hook that returns dependencies for retrieving an existing value of a choice. This is
+     * useful for things like hooking into an existing cache (eg. [useViewModelCache](doc:useViewModelCache])). The value
+     * returned here is passed as the second parameter to `retrieve`. In addition the `existingValues` key is
+     * passed through to `useAsyncValue` as the list of items it can resolve existing values from.
+     *
+     * What this function is passed depends on the implementation but when used with
+     * [useAsyncChoices](doc:useAsyncChoices) it will be passed `id` if there's a current value
+     * and it's not an array, `ids` if there's a current value and it is an array, `existingValues` which is
+     * the values returned by `list` (may be null if `list` not yet called) and `retrieveOptions`
+     * (any additional options passed on the `retrieveOptions` prop to `useAsyncChoices`).
+     * Official presto widgets all use `useAsyncChoices` and so will use these parameters.
+     */
+    useRetrieveProps(args: any): any;
+
+    /**
+     * Function to resolve specific values. This is used to know how to render the label for a value(s).
+     *
+     * The first parameter is the value to retrieve (will be an array when `multiple` is true).
+     *
+     * `deps` is the value returned by `useRetrieveProps`.
+     */
+    retrieve(
+        value: ValueType[] | ValueType,
+        deps?: any
+    ): Promise<typeof value extends ValueType[] ? ItemType[] : ItemType>;
+
+    /**
+     * Generate the list of choices. This can return an array of single choices or grouped choices.
+     *
+     * A grouped choice is a 2 element Array with a label and a list of choices.
+     *
+     * @returns Either an array of single choices or grouped choices.
+     */
+    getChoices(items: ItemType[]): (Choice<ValueType> | ChoicesGrouped<ValueType>)[];
+
+    /**
+     * Get a label for an item
+     */
+    getLabel(item: ItemType): React.ReactNode;
+
+    /**
+     * Return label to use when an item can't be found. This can be used by widgets to control
+     * what is rendered when an item for a value cannot be found (eg. when it's deleted or
+     * when it's loading). The exact details of how this is used depend on the widget.
+     */
+    getMissingLabel(value: ValueType): React.ReactNode;
```

Comment:
Is this intended to handle `null` values too or would the Widget handle that?

Reply:
No this is just for if a value is present but no label can be found

Reply:
(widget handles it - there's a default on the widget or can be specified as a prop)